### DISCUSSION TITLE
How to calculate the pixel count of DXF units for a rendered PNG file?

### DISCUSSION THREAD

**Discussion by yingma111** [mozman/ezdxf#228] [Q&A] [answered]
I do dxf file to png file, and now I want to kown how many pixels in png file 1 millimeter in dxf file is equal to?

**Comment by yingma111** [comment]
My code is same with [(https://github.com/mozman/ezdxf/issues/226)].
Please reply me soon.
Super thanks.

**Comment by mbway** [comment, marked as answer]
fundamentally, you need the transform between the CAD coordinate system and the image coordinate system. Then, assuming the cad coordinates are in millimetres anyway, any distance in the CAD can be converted to a distance in pixels and vice versa. If the CAD isn't in millimetres, then you need to first scale between the CAD units and millimetres.

see my answer here: https://github.com/mozman/ezdxf/issues/219#issuecomment-681132750 for one possible solution for getting the transformation between the cad coordinate system and pixels of the resulting image.
note that to transform a vector (i.e. distance) rather than a point using a homogeneous transformation matrix can be done with:
```python
def transform_vector(x:float, y: float, m: np.ndarray) -> Tuple[float, float]:
    return tuple((m @ [x, y, 0])[:2].tolist())
```

@mozman, since multiple people have found the need to obtain this transformation, do you think some facility for this could be added to ezdxf directly? When I wrote the solution to #219 I used numpy arrays rather than Matrix44 since I needed a homogeneous matrix to transform 2D points (i.e. 3x3 matrix). I suppose a Matrix44 could be used which (orthographically) projects 3D points down to z=0 for the image coordinate system. What do you think?
Also, possibly more testing of my solution might be required. I found that my solution was pretty much correct but I'm not sure about sub-pixel accuracy. It's a bit tricky since the CAD coordinates are continuous but the image coordinate system is discrete (with pixels indexed from the top left) and also flipped in y w.r.t. the cad coordinate system.

**Comment by mbway** [comment]
also, if you just want a single scale factor, it should be sufficient in your case to take the ratio between the x axis limits (i.e. the data width) and the image width. So
```python
x1, x2 = ax.get_xlim()
width_mm = x2 - x1  # assuming cad coordinates are in mm
width_px = # <obtain image width>
mm_to_px = width_px / width_mm

length_mm = 10
length_px = length_mm * mm_to_px
```

**Comment by mozman** [comment]
DXF and measurement units are not a good match, there is practically no documentation, therefore I don't want to open this can of worms ;-). But this is good topic for a How-To, I do the "restructured text" formatting if you provide the content.

**Comment by yingma111** [comment]
I do test on 284 files and results show me that pix/mm values are right

**Comment by yingma111** [comment]
Super thanks mbway

**Comment by Aidmx** [comment]
when coordinate of entities in dxf , i found some  lines offset a little to pixel on png.

**Comment by mbway** [comment]
@Aidmx are you saying that simply using the scale factor to get the transformation between CAD and the png results in a slight error in the resulting pixel coordinates? If so then I also encountered this.

I think it's to do with the coordinate systems being flipped relative to one another. The CAD coordinate system has +y pointing up, and images have +y pointing down. This also means that a pixel coordinate refers to the top left of a pixel whereas simply scaling a CAD coordinate to the image size will be addressing the bottom left of a pixel. I think this is where the small error comes from.

My solution here: https://github.com/mozman/ezdxf/issues/219#issuecomment-681132750 adds 1 to the coordinates before flipping and gave more accurate results for me, but the algorithm may not be perfect.
